packages/react-devtools-shared/src/devtools/views/Components/InspectedElement.js

```diff
@@ -222,10 +224,21 @@ export default function InspectedElementWrapper(_: Props) {
       return;
     }
 
-    const url = new URL(editorURL);
-    url.href = url.href.replace('{path}', source.fileName);
-    url.href = url.href.replace('{line}', String(source.lineNumber));
-    window.open(url);
+    let url = editorURL;
+    url = url.replace('{path}', source.fileName);
+    url = url.replace('{line}', String(source.lineNumber));
+    if (url.startsWith('http')) {
+      // fix https://github.com/facebook/react/issues/24795
+      fetch(url);
+    } else {
+      // fix https://github.com/facebook/react/issues/24731
+      chrome.runtime.sendMessage({
+        payload: {
+          type: 'openInEditorByNewTab',
+          url,
+        },
+      });
```

Comment:
Can we actually do this without sending the message to background content script?

I've tried to use `chrome.tabs.create` from `InspectedElement.js` component directly and it works.

Reply:
Sorry, you can check the workflow. If you use chrome.tabs directly, an error will be reported. It has been too long since then and I should have encountered this situation at that time, so I used sendMessage.

Reply:
Running Flow on the dom-node renderer...
Error --------------------------- packages/react-devtools-shared/src/devtools/views/Components/InspectedElement.js:229:7

Cannot resolve name `chrome`. [cannot-resolve-name]

   229|       chrome.tabs.create({
              ^^^^^^



Found 1 error
Flow failed for the dom-node renderer


Exited with code exit status 2
CircleCI received exit code 2

Reply:
@hoxyq Hello, what do I need to do now?

Reply:
> @hoxyq Hello, what do I need to do now?

Hey, sorry for the delay. I just figured out that these changes will probably break the standalone build of the DevTools that we have. We wrap these interfaces in electron application, and from what I can see, it [does not support](https://www.electronjs.org/docs/latest/api/extensions#chrometabs) `chrome.tabs.create`.

I guess we need to go with your initial solution, where we will send an event through bridge (not chrome.runtime) and then will subscribe to it:
- For standalone in [`react-devtools-core/src/standalone.js](https://github.com/facebook/react/blob/ee4233bdbc71a7e09395a613c7dde01194d2a830/packages/react-devtools-core/src/standalone.js)
- For extension in [`react-devtools-extensions/src/main.js](https://github.com/facebook/react/blob/ee4233bdbc71a7e09395a613c7dde01194d2a830/packages/react-devtools-extensions/src/main.js)

Based on the surface (standalone and extension), we should implement the event handling accordingly, I think for standalone we still should use `window.open` one.

Let me know if you will have any questions, on how to test it, run it or setting up the standalone build.

Reply:
@hoxyq II don't have any further questions, so my understanding is that I should revert back to the initial code, which sends an event through bridge. Additionally, I should remove that fetch, right? Is my understanding correct?

Reply:
> @hoxyq II don't have any further questions, so my understanding is that I should revert back to the initial code, which sends an event through bridge. Additionally, I should remove that fetch, right? Is my understanding correct?

Almost, your initial version was using `chrome.runtime` to send an event, but you need to send it through bridge here, just because this is the common way in the project.

Also, in your initial version you were subscribing to this event in `background.js`, but because we sending an event through bridge, you should now do this in `main.js`, where we are initializing it.

`main.js` file is included only in extension build, so for standalone version to handle this new event, you need to add a corresponding listener in [`react-devtools-core/src/standalone.js`](https://github.com/facebook/react/blob/ee4233bdbc71a7e09395a613c7dde01194d2a830/packages/react-devtools-core/src/standalone.js)

Reply:
Okay, I'll give it a try.

Reply:
I'm really sorry, I've been really busy lately. I will definitely finish the revision this week.

Reply:
I'm really sorry, I've been a bit busy these days. I will finish the revision this week.

Reply:
> I'm really sorry, I've been a bit busy these days. I will finish the revision this week.

No worries, there is no rush. Once updated, I will come back to review the changes once again. Feel free to ask any questions that might occur

Reply:
@hoxyq 
I execute `chrome.tabs.create` inside `src/main.js`, it will throw a error that `chrome.tabs is undefined`. I tried using `chrome.devtools.inspectedWindow.eval(`window.open(${url});`)` to open a new tab, but it requires selecting "Allow" when Firefox blocks it for the first time.

I can use `chrome.devtools.inspectedWindow.eval(window.open(${url});)` to solve the problem, or continue to look for the correct way to call `chrome.tabs.create`.

Reply:
@hoxyq I see that chrome.devtools.inspectedWindow.eval is a very stable way to use in the system.

Reply:
Previously, we used to directly use window.open in the DevTools page. I think a smoother way to fix this might be to use chrome.devtools.inspectedWindow.eval(window.open('${url}');).

Reply:
> 

I think the right way to do it is:
1. Call `chrome.tabs.create` only from `background.js`, which is a background script. The problem with `main.js` is that it is a content script, which has a restricted list of APIs.
2. `main.js` and `background.js` can communicate with each other via `chrome.runtime`.
3. Make sure you added `tabs` permission to Firefox's manifest.

Reply:
> > 
> 
> I think the right way to do it is:
> 
> 1. Call `chrome.tabs.create` only from `background.js`, which is a background script. The problem with `main.js` is that it is a content script, which has a restricted list of APIs.
> 2. `main.js` and `background.js` can communicate with each other via `chrome.runtime`.
> 3. Make sure you added `tabs` permission to Firefox's manifest.

But still, we will need to use `window.open` for standalone version, because I believe `background.js` script is not shipped there.

Sorry with bringing some confusion with `main.js`

Reply:
I have already added the necessary tabs permission to Firefox's manifest, but it still doesn't work. 

Then we can add a conditional statement: if it is Chrome, use `window.open` directly; if it is Firefox, then call `chrome.devtools.inspectedWindow.eval(window.open('${url}');)`

Reply:
I have already added the necessary tabs permission to Firefox's manifest, but it still doesn't work.

Then we can add a conditional statement: if it is Chrome, use window.open directly; if it is Firefox, then call chrome.devtools.inspectedWindow.eval(window.open('${url}');)

Reply:
If we handle it this way, then you wouldn't need to adjust the `background.js` file.

Reply:
> If we handle it this way, then you wouldn't need to adjust the `background.js` file.

1. `window.open` should be used only for standalone version of the DevTools, all browser extensions (Chrome, Firefox, Edge) should use `chrome.tabs` api.
2. You should not use browser apis (`chrome.tabs` or others) inside view components, they should be platform-agnostic, because they are used for both browser environment and standalone application (limited Chromium environment)

Because of point 2 you need to propagate these events through `bridge` or `agent` (which are platform agnostic and used on every of them) and then subscribe to them in:
1. `background.js` (extension-specific)
2. `react-devtools-core/src/standalone.js` - I believe this is the right one for standalone version
3. `react-devtools-inline/src/frontend.js` - for browser-inline version (doesn't have access to extension apis like `chrome.tabs`, so we need to use `window.open` here)